File: generated_code/GeminiPro/Java/001_SplayTree.java
/**
 * A complete, self-contained Java class that implements an integer set using a Splay Tree.
 *
 * A splay tree is a self-balancing binary search tree with the additional
 * property that recently accessed elements are quick to access again.
 * It performs basic operations such as insertion, look-up, and removal
 * in O(log n) amortized time. All core operations (insert, delete, search)
 * perform a "splay" operation, which moves the accessed (or inserted/deleted)
 * node to the root of the tree.
 */
public class SplayTree {

    /**
     * Represents a node in the Splay Tree. This is a private inner class
     * as it's an implementation detail of the SplayTree.
     */
    private class Node {
        int key;
        Node left, right;

        /**
         * Constructs a new node with the given key.
         * @param key The integer key for the node.
         */
        public Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        root = null;
    }

    /**
     * Performs a right rotation on the subtree rooted at x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the rotated subtree.
     */
    private Node rightRotate(Node x) {
        Node y = x.left;
        x.left = y.right;
        y.right = x;
        return y;
    }

    /**
     * Performs a left rotation on the subtree rooted at x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the rotated subtree.
     */
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * Splays the node with the given key to the root of the tree.
     * If the key is not found, the last accessed node on the search path is splayed to the root.
     * This is a recursive, bottom-up splaying implementation.
     *
     * @param h The root of the current subtree.
     * @param key The key to splay towards the root.
     * @return The new root of the splayed subtree.
     */
    private Node splay(Node h, int key) {
        if (h == null || h.key == key) {
            return h;
        }

        if (key < h.key) {
            // Key is in the left subtree
            if (h.left == null) {
                // Key not found, return the last accessed node
                return h;
            }
            if (key < h.left.key) { // Zig-Zig (Left-Left)
                h.left.left = splay(h.left.left, key);
                h = rightRotate(h);
            } else if (key > h.left.key) { // Zig-Zag (Left-Right)
                h.left.right = splay(h.left.right, key);
                if (h.left.right != null) {
                    h.left = leftRotate(h.left);
                }
            }
            // Second rotation for Zig-Zig or the only rotation for Zig
            return (h.left == null) ? h : rightRotate(h);
        } else { // key > h.key
            // Key is in the right subtree
            if (h.right == null) {
                // Key not found, return the last accessed node
                return h;
            }
            if (key > h.right.key) { // Zig-Zig (Right-Right)
                h.right.right = splay(h.right.right, key);
                h = leftRotate(h);
            } else if (key < h.right.key) { // Zig-Zag (Right-Left)
                h.right.left = splay(h.right.left, key);
                if (h.right.left != null) {
                    h.right = rightRotate(h.right);
                }
            }
            // Second rotation for Zig-Zig or the only rotation for Zig
            return (h.right == null) ? h : leftRotate(h);
        }
    }

    /**
     * Searches for a key in the tree. If the key is found, it is splayed
     * to the root, and the key is returned. If not found, the last-accessed
     * node is splayed to the root, and null is returned.
     *
     * @param key The key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        root = splay(root, key);
        if (root != null && root.key == key) {
            return key;
        }
        return null;
    }

    /**
     * Inserts a key into the splay tree.
     * If the key already exists, the tree is splayed on that key but not modified.
     *
     * @param key The key to insert.
     */
    public void insert(int key) {
        // If tree is empty, create a new root
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree to bring the closest node to the root
        root = splay(root, key);

        // If key is already present, do nothing (it's a set)
        if (root.key == key) {
            return;
        }

        // Otherwise, create the new node and make it the root
        Node newNode = new Node(key);
        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        } else { // key > root.key
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        root = newNode;
    }

    /**
     * Deletes a key from the splay tree.
     * If the key is not found, the tree is splayed on the last-accessed node
     * but not modified.
     *
     * @param key The key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // Splay the tree for the key
        root = splay(root, key);

        // If key is not in the tree, do nothing
        if (root.key != key) {
            return;
        }

        // Now, the node to be deleted is at the root.
        // We need to join its left and right subtrees.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            // If there's no left subtree, the right subtree becomes the new tree
            root = rightSubtree;
        } else {
            // Splay the left subtree to bring its maximum element to its root.
            // This max element will become the new root of the combined tree.
            // Splaying for a key larger than any in the subtree (like the deleted key)
            // achieves this.
            Node newRoot = splay(leftSubtree, key);

            // The new root (max of left subtree) has no right child, so
            // we can attach the original right subtree there.
            newRoot.right = rightSubtree;
            root = newRoot;
        }
    }
}

File: generated_code/GeminiPro/Python/001_SplayTree.py
import sys

class SplayTree:
    """
    A self-contained Python class implementing a Splay Tree.

    This class provides a dictionary-like set for storing unique integer keys.
    It supports insert, delete, and search operations. The key feature is the
    splaying operation performed on each access, which moves the accessed
    node (or its parent if the node is not found) to the root of the tree.
    This amortizes the cost of operations to O(log n).
    """

    class _Node:
        """
        A private inner class representing a node in the Splay Tree.
        """
        def __init__(self, key):
            self.key = key
            self.parent = None
            self.left = None
            self.right = None

    def __init__(self):
        """
        Initializes an empty Splay Tree.
        """
        self.root = None

    def _left_rotate(self, x):
        """
        Performs a left rotation on the given node x.
        """
        y = x.right
        if y:
            x.right = y.left
            if y.left:
                y.left.parent = x
            y.parent = x.parent
        
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        
        if y:
            y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """
        Performs a right rotation on the given node x.
        """
        y = x.left
        if y:
            x.left = y.right
            if y.right:
                y.right.parent = x
            y.parent = x.parent

        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
            
        if y:
            y.right = x
        x.parent = y

    def _splay(self, node):
        """
        Performs the splaying operation on the given node, moving it to the root.
        """
        while node.parent:
            parent = node.parent
            grandparent = parent.parent
            if not grandparent:
                # Zig case
                if node == parent.left:
                    self._right_rotate(parent)
                else:
                    self._left_rotate(parent)
            elif node == parent.left and parent == grandparent.left:
                # Zig-Zig case (left-left)
                self._right_rotate(grandparent)
                self._right_rotate(parent)
            elif node == parent.right and parent == grandparent.right:
                # Zig-Zig case (right-right)
                self._left_rotate(grandparent)
                self._left_rotate(parent)
            elif node == parent.right and parent == grandparent.left:
                # Zig-Zag case (left-right)
                self._left_rotate(parent)
                self._right_rotate(grandparent)
            else:
                # Zig-Zag case (right-left)
                self._right_rotate(parent)
                self._left_rotate(grandparent)

    def search(self, key):
        """
        Searches for a key in the tree.

        If the key is found, the corresponding node is splayed to the root
        and True is returned.
        If the key is not found, the last accessed node (the would-be parent)
        is splayed to the root and False is returned.

        Args:
            key (int): The integer key to search for.

        Returns:
            bool: True if the key is found, False otherwise.
        """
        last_visited = None
        current = self.root
        while current:
            last_visited = current
            if key < current.key:
                current = current.left
            elif key > current.key:
                current = current.right
            else:
                # Key found, splay the node and return True
                self._splay(current)
                return True
        
        # Key not found, splay the last visited node (parent) if it exists
        if last_visited:
            self._splay(last_visited)
        
        return False

    def insert(self, key):
        """
        Inserts a key into the tree.

        If the key already exists, the existing node is splayed to the root.
        If the key is new, it is inserted and the new node is splayed to the root.

        Args:
            key (int): The integer key to insert.
        """
        if not self.root:
            self.root = self._Node(key)
            return

        current = self.root
        parent = None
        while current:
            parent = current
            if key < current.key:
                current = current.left
            elif key > current.key:
                current = current.right
            else:
                # Key already exists, splay the node and return
                self._splay(current)
                return
        
        # Insert the new node
        new_node = self._Node(key)
        new_node.parent = parent
        if key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node
        
        # Splay the new node to the root
        self._splay(new_node)

    def delete(self, key):
        """
        Deletes a key from the tree.

        The operation first searches for the key, which splays the node (or its
        parent) to the root. If the key is found at the root, it is removed,
        and the remaining subtrees are joined.

        Args:
            key (int): The integer key to delete.
        """
        # Search for the key. This will splay the node or its parent to the root.
        if not self.search(key):
            # Key not in the tree. search() already splayed the would-be parent.
            return

        # After search, if the key was found, it is now at the root.
        node_to_delete = self.root
        
        left_subtree = node_to_delete.left
        right_subtree = node_to_delete.right

        if not left_subtree:
            # No left child, the right subtree becomes the new tree.
            self.root = right_subtree
            if right_subtree:
                right_subtree.parent = None
        elif not right_subtree:
            # No right child, the left subtree becomes the new tree.
            self.root = left_subtree
            if left_subtree:
                left_subtree.parent = None
        else:
            # Both children exist.
            # 1. Detach the left subtree.
            left_subtree.parent = None
            
            # 2. Find the maximum node in the left subtree.
            max_in_left = left_subtree
            while max_in_left.right:
                max_in_left = max_in_left.right
            
            # 3. Splay this maximum node to the root of the left subtree.
            #    Temporarily make the left subtree the main tree to use _splay.
            self.root = left_subtree
            self._splay(max_in_left)
            
            # 4. The new root (max_in_left) now has the right_subtree as its
            #    right child. By property, it has no right child before this.
            self.root.right = right_subtree
            right_subtree.parent = self.root